samza-yarn/src/test/java/org/apache/samza/job/yarn/util/hadoop/TestHttpFileSystem.java

```diff
@@ -0,0 +1,156 @@
+* to you under the Apache License, Version 2.0 (the
+* "License"); you may not use this file except in compliance
+* with the License.  You may obtain a copy of the License at
+*
+*   http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing,
+* software distributed under the License is distributed on an
+* "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+* KIND, either express or implied.  See the License for the
+* specific language governing permissions and limitations
+* under the License.
+*/
+
+package org.apache.samza.job.yarn.util.hadoop;
+
+import junit.framework.Assert;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FSDataInputStream;
+import org.apache.hadoop.fs.Path;
+import org.apache.samza.coordinator.server.HttpServer;
+import org.apache.samza.util.hadoop.HttpFileSystem;
+import org.eclipse.jetty.servlet.DefaultServlet;
+import org.eclipse.jetty.servlet.ServletHolder;
+import org.junit.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.ByteArrayInputStream;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.SocketTimeoutException;
+import java.net.URI;
+import java.util.concurrent.CountDownLatch;
+
+/**
+ * Test the behavior of {@link HttpFileSystem}
+ */
+public class TestHttpFileSystem {
+
+  private static final Logger LOG = LoggerFactory.getLogger(TestHttpFileSystem.class);
+  /**
+   * Number of bytes the server should stream before hanging the TCP connection.
+   */
+  private static final int THRESHOLD_BYTES = 5;
+  private static final String RESPONSE_STR = "HELLO WORLD";
+
+  private final CountDownLatch serverWaitLatch = new CountDownLatch(1);
+
+  private Exception clientException;
+  private Exception serverException;
+
+  /**
+   * A {@link HttpServlet} implementation that streams its response to the client one byte at a time.
+   */
+  private class PartialFileFetchServlet extends HttpServlet {
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) {
+
+      //Mimic download of the package tar-ball
+      response.setContentType("application/gzip");
+      response.setStatus(HttpServletResponse.SC_OK);
+
+      try {
+        int currByteWritten = -1;
+        int numBytesWritten = 0;
+        //Begin streaming a response.
+        InputStream in = new ByteArrayInputStream(RESPONSE_STR.getBytes());
+        OutputStream out = response.getOutputStream();
+
+        while ((currByteWritten = in.read()) != -1) {
+          out.write(currByteWritten);
+          out.flush();
+          numBytesWritten++;
+
+          //Hang the connection until the read timeout expires on the client side.
+          if (numBytesWritten >= THRESHOLD_BYTES) {
+            serverWaitLatch.await();
```

Comment:
Added a return to be more defensive. In my tests, it did not cause any exceptions (because the socket is technically not closed until the process is killed).